[AutoDiff] Support differentiation of loops.

dan-zheng opened the pull request:
- Change predecessor enum generation to support loops.
  - Generate indirect predecessor enums for BBs in loops.
  - Handle boxed payloads of indirect enums.
- Traverse basic blocks in post-order post-dominance order.
  - This is necessary for computational correctness.
- Add loop tests (`for-in`, `while`, `repeat-while`, `break`/`continue`, nested).

---

Examples:
```swift
func for_loop(_ x: Float) -> Float {
  var result = x
  for _ in 1..<3 {
    result = result * x
  }
  return result
}
expectEqual((8, 12), valueWithGradient(at: 2, in: for_loop))
expectEqual((27, 27), valueWithGradient(at: 3, in: for_loop))
```
```swift
func nested_loop(_ x: Float, count: Int) -> Float {
  var outer = x
  for _ in 1..<count {
    outer = outer * x

    var inner = outer
    var i = 1
    while i < count {
      inner = inner + x
      i += 1
    }
    outer = inner
  }
  return outer
}
expectEqual((20, 22), valueWithGradient(at: 2, in: { x in nested_loop(x, count: 3) }))
expectEqual((104, 66), valueWithGradient(at: 4, in: { x in nested_loop(x, count: 3) }))
```

[Verified loop differentiation correctness](https://gist.github.com/dan-zheng/d153d950687820418dbae65517259ff0) against PyTorch and Tangent.

Exposed [TF-584](https://bugs.swift.org/browse/TF-584): derivative computation bug for repeat-while loops.
Exposed [TF-585](https://bugs.swift.org/browse/TF-585): AllocBoxToStack crash with `-O` for nested loop AD.

dan-zheng committed:
[AutoDiff] Support differentiation of loops.

- Change predecessor enum generation to support loops.
  - Generate indirect predecessor enums for BBs in loops.
  - Handle boxed payloads of indirect enums.
- Traverse basic blocks in post-order post-dominance order.
  - This is necessary for computational correctness.
- Add loop tests (`for-in`, `while`, nested).

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3055,15 +3042,28 @@ class VJPEmitter final
+    if (!boxType)
+      return builder.createEnum(loc, pbStructVal, enumEltDecl, enumLoweredTy);
+    // Otherwise, box the pullback struct value and create an enum.
+    auto *allocBox = builder.createAllocBox(loc, boxType);
```
This `alloc_box` is not being deallocated and may cause leaks - to be verified.
Loop AD leak checking tests are added in `test/AutoDiff/leakchecking.swift`.

dan-zheng replied:
Resolved - see [this comment](https://github.com/apple/swift/pull/25558#pullrequestreview-250897375). Deallocation is handled automatically during `release_value` for pullback structs.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3738,16 +3741,17 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+  /// Mapping from original basic blocks to local adjoint values to be cleaned
+  /// up. This is populated when adjoint emission is run on one basic block and
+  /// cleaned before processing another basic block.
+  DenseMap<SILBasicBlock *, SmallVector<AdjointValue, 8>>
```
Note: changed `blockLocalAdjointValues` to a `DenseMap` for more principled/predictable behavior.
Previously, using `SmallVector` caused bugs when the original entry BB isn't the last visited BB during adjoint generation.

rxwei commented on test/AutoDiff/control_flow.swift:
```diff
@@ -519,4 +519,46 @@ ControlFlowTests.test("Enums") {
+  expectEqual((8, 12), valueWithGradient(at: 2, in: for_loop))
+  expectEqual((27, 27), valueWithGradient(at: 3, in: for_loop))
+
+  func while_loop(_ x: Float) -> Float {
```
Please also test a `repeat`-`while` loop.

dan-zheng replied:
Added tests for `repeat-while` loops and `continue`/`break` statements in https://github.com/apple/swift/pull/25558/commits/32f66294d96791f67f15234948b493b76b9d31da.
Exposed [TF-584](https://bugs.swift.org/browse/TF-584): derivative computation bug for repeat-while loops. Thanks for pointing this out!

rxwei reviewed (commented):
~~I couldn't find any emission of `dealloc_box` and this is apparently leaking a lot more memory (all pullback structs in loops are kept alive). This needs to be fixed before merging.~~
Actually, this should be handled automatically when you `release_value` the pullback struct.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1737,6 +1736,7 @@ static void dumpActivityInfo(SILFunction &fn,
     for (auto &inst : bb)
       for (auto res : inst.getResults())
         dumpActivityInfo(res, indices, activityInfo, s);
+    s << "\n";
```
```suggestion
    s << '\n';
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/25558/commits/32f66294d96791f67f15234948b493b76b9d31da.

rxwei reviewed (approved):
The patch is pretty clean. LGTM!

dan-zheng committed:
Add tests.

Add tests for repeat-while loops and break/continue statements.
Expose TF-584: incorrect derivative computation for repeat-while loops.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng committed:
Minor fixes.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
Two tests fail only with optimizations, which I hadn't tested locally:
```
Failing Tests (2):
    Swift(linux-x86_64) :: AutoDiff/leakchecking.swift
    Swift(linux-x86_64) :: AutoDiff/control_flow.swift
```

```
bool swift::StackNesting::solve(): Assertion `!(BI.Block->getTerminator()->isFunctionExiting() && Bits.any()) && "stack location is missing dealloc"' failed.
```
Perhaps there are some missing deallocations in generated code. Will investigate later.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3055,15 +3042,31 @@ class VJPEmitter final
+                        getBufferSOQ(projectBox->getType().getASTType(), *vjp));
+    // NOTE(TF-585): `fix_lifetime` is generated to avoid AllocBoxToStack crash
+    // for nested loop AD.
+    // builder.createFixLifetime(loc, allocBox);
```
And yet it's commented out :)

dan-zheng replied:
Thanks for the catch c:

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Avoid AllocBoxToStack crash (TF-585).

Generate `fix_lifetime` instruction on `alloc_box` to avoid crash.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
